README

Contents
--------

0.) Name
1.) Motivation
2.) inotify vs fanotify
3.) Installing
4.) How to use
5.) Example of usage
6.) Recommended configuration
7.) Known issues
8.) Support
9.) Developing

                 _____________________________________________                


0.) Name
--------

Why "clsync"? The first name of the utility was "insync" (due to inotify), but
then I suggested to use "fanotify" instead of "inotify" and utility was been
renamed to "fasync". After that I started to intensively write the program.
However I faced with some problems in "fanotify", so I was have to temporary
fallback to "inotify", then I decided, that the best name is "Runtime Sync" or
"Live Sync", but "rtsync" is a name of some corporation, and "lsync" is busy
by "lsyncd" project ("https://github.com/axkibe/lsyncd"). So I called it
"clsync", that should be interpreted as "lsync, but on c" due to "lsyncd" that
written on "LUA" and may be used for the same purposes.

                 _____________________________________________                


1.) Motivation
--------------

This utility was been writted for two purposes:
	- for making failover clusters
	- for making backups of them

To do failover cluster I've tried a lot of different solutions, like "simple 
rsync by cron", "glusterfs", "ocfs2 over drbd", "common mirrorable external 
storage", "incron + perl + rsync", "inosync", "lsyncd" and so on. Currently we
are using "lsyncd", "ceph" and "ocfs2 over drbd". However all of this
solutions doesn't arrange me, so I was have to write own utility for this
purpose.

To do backups we also tried a lot of different solution, and again I was have
to write own utility for this purpose.

The best known (for me) replacement for this utility is "lsyncd", however:
	- It's code is >½ on LUA. There a lot of problems connected with it,
for example:
	    - It's more difficult to maintain the code with ordinary sysadmin.
	    - It really eats 100% CPU sometimes.
	    - It requires LUA libs, that cannot be easily installed to few
of our systems.
	- It's a little buggy. That may be easily fixed for our cases,
but LUA. :(
	- It doesn't support pthread or something like that. It's necessary
to serve huge directories with a lot of containers right.
	- It cannot run rsync for a pack of files. It runs rsync for every
event. :(
	- Sometimes, it's too complex in configuration for our situation.
	- It can't set another event-collecting delay for big files. We don't
want to sync big files (>1GiB) so often as ordinary files.

Sorry, if I'm wrong. Let me know if it is, please :). "lsyncd" - is really
good and useful utility, just it's not appropriate for us.

                 _____________________________________________                


2.) inotify vs fanotify:
------------------------

It's said, that fanotify is much better, than inotify. So I started to write 
this program with using of fanotify. However I encountered the problem, that
fanotify was unable to catch some important events at the moment of writing
the program, like "directory creation" or "file deletion". So I switched to
"inotify", leaving the code for "fanotify" in the safety... So, don't use
"fanotify" in this utility ;).

                 _____________________________________________                


3.) Installing
--------------

First of all, you should install dependencies to compile clsync. As you can
see from GNUmakefile clsync depends only on "glib-2.0", so on debian-like
systems you should execute something like "apt-get install libglib2.0-dev".

Next step is compiling. To compile usually it's enough to execute "make".

Next step in installing. To install usually it's enough to execute
"su -c 'make install'".

Also, debian-users can use my repository to install the clsync:
deb [arch=amd64] http://mirror.mephi.ru/debian-mephi/ unstable main

                 _____________________________________________                


4.) How to use
--------------

How to use is described in "man" ;). What is not described, you can ask me
personally (see "Support").

                 _____________________________________________                


5.) Example of usage
--------------------

Example of usage, that works on my PC is in directory "example". Just run
"clsync-start.sh" and try to create/modify/delete files/dirs in
"example/testdir/from". All modifications should appear (with some delay) in
directory "example/testdir/to" ;)

                 _____________________________________________                


6.) Recommended configuration
-----------------------------

First of all, recommended and not recommended options are notices in the
manpage.

However let's describe 3 situations:

- You're backing-up over very slow channel:
clsync -l backup -R -d /dev/shm -t 600 -T 3600 -B $[1024 * 1024 * 16] /home/user /home/clsync/bin/clsync-actionscript.sh /home/clsync/clsync-rules

This will minimize network traffic. And pthread-ing is removed due to rarely
updating.

- You're syncing ordinary web-server over 1Gbs channel:
clsync -l mirror -p -R -d /dev/shm /var/www /home/clsync/bin/clsync-actionscript.sh /home/clsync/clsync-rules

- You're syncing only few files from huge file tree (with a great lot of
excludes):
clsync -l mirror -p -R -d /dev/shm -I /home/user /home/clsync/bin/clsync-actionscript.sh /home/clsync/clsync-rules

                 _____________________________________________                


7.) Known issues
----------------

    1.) Doesn't compiles on old systems

	In this case you should compile with command "make onoldsystem"

    2.) Exits with error "Error: Cannot inotify_add_watch() on [...]"

	In this case you should increase "fs.inotify.max_user_watches" value,
	for example with command: sysctl fs.inotify.max_user_watches=26214400

                 _____________________________________________                

8.) Support
-----------

    To get support, you can contact with me this ways:
	- IRC: SSL+UTF-8 irc.campus.mephi.ru:6695#mephi,xaionaro,xai
	- e-mail: <xai@mephi.ru>, <dyokunev@ut.mephi.ru>, <xaionaro@gmail.com>
		PGP pubkey: 0x8E30679C

                 _____________________________________________                

9.) Developing
-----------

    I started to write "DEVELOPING" file. You can look there if you wish. ;)

    I'll be glad to receive code contribution :)

                 _____________________________________________                

				-- Dmitry Yu Okunev <xai@mephi.ru> 0x8E30679C

P.S.: Sorry for my English and for the code :)



debian/changelog
clsync (0.4.5-2) unstable; urgency=medium

  [ Dmitrii Okunev ]
  * Remove dependency on libcgroup-dev which is no longer functional
    - a stub using libcgroup2 might be usable, once it spins up
    - See bugs.debian.org/959022 for details

  [ Barak A. Pearlmutter ]
  * Bump policy

 -- Barak A. Pearlmutter <bap@debian.org>  Thu, 18 Mar 2021 10:26:01 +0000

clsync (0.4.5-1) unstable; urgency=medium

  [ Andrew Savchenko]
  * New upstream relese
  * Added --sync-on-quit support
  * Added $TMPDIR support
  * Multiple bug fixes

  [Dmitrii Okunev]
  * Added systemd unit-file

  [ Barak A. Pearlmutter ]
  * Depend on newer debhelper (>= 9.20160709) rather than dh-systemd.
  * Set upstream metadata fields: Bug-Database, Bug-Submit, Repository,
    Repository-Browse.
  * Update standards version to 4.5.0, no changes needed.
  * Merge upstream fix for symbols pthread_mutex_init_shared
    pthread_mutex_reltimedlock pthread_mutex_destroy_shared in
    libclsync.so.0.0.0 not found in any libraries.
  * Bump to debhelper 13.
  * Address lintian complaint about init.d file
  * Git ignore some debris
  * Tell dh_missing not to worry about the *.la file

 -- Barak A. Pearlmutter <bap@debian.org>  Mon, 09 Nov 2020 16:35:16 +0000

clsync (0.4.3-1) unstable; urgency=medium

  [ Dmitry Yu Okunev ]
  * Fixed mode "simple": use absolute paths
  * Better error reporting on bad arguments/configuration.
  * Disabled "seccomp" features to make the package buildable
    on non-x86 platforms (closes: #844778)
  * Updated the manpage

  [ Barak A. Pearlmutter ]
  * New Upstream Version
  * Bump policy version
  * Bump to debhelper 12
  * Update uscan debian/watch file
  * add git-buildpackage configuration file
  * dependency per lintian tag skip-systemd-native-flag-missing-pre-depends
  * move library package to section: libs

 -- Barak A. Pearlmutter <bap@debian.org>  Wed, 08 May 2019 12:37:57 +0100

clsync (0.4.2-1) unstable; urgency=medium

  [ Dmitry Yu Okunev ]
  * Fixed building on Debian/kFreeBSD.
  * Fixed "Permission denied" caused by "--enable-debug=no".
  * Option "check-execvp-args" renamed to "check-execvp-arguments".
  * Fixed building and running on i386 with enabled seccomp.
  * Fixed a "Segmentation fault" and a deadlock on high load with
    enabled "thread splitting".
  * Forbidded to enable together: "--theading" and "--seccomp-filter"
    (bug: https://github.com/xaionaro/clsync/issues/139).
  * Forbidded option "--debug" if compiled with "--enabled-debug=no"
  * Added rule types: "m", "s" and "W"
  * Fixed a bug related to sync retries

  [ Barak A. Pearlmutter ]
  * New upstream release
  * Bump standards version
  * Fix some typos
  * Harden

 -- Barak A. Pearlmutter <bap@debian.org>  Wed, 05 Oct 2016 11:31:46 +0100

clsync (0.4.1-1) unstable; urgency=medium

  * New upstream release with a lot of fixes

 -- Barak A. Pearlmutter <bap@debian.org>  Tue, 12 May 2015 08:53:54 +0100

clsync (0.4-1) unstable; urgency=medium

  [ Dmitry Yu Okunev ]
  * A lot of fixes

  [ Barak A. Pearlmutter ]
  * Merge upstream, in particular debian/ tweaks
  * don't install compile-time stuff that shouldn't be
  * bump debian standards version, no changes necessary
  * update debian/watch file, mangle upstream -rc
  * update .gitignore
  * correct some spelling errors
  * man dash hyphen slash patch

 -- Barak A. Pearlmutter <bap@debian.org>  Fri, 08 May 2015 10:43:41 +0100

clsync (0.3-1) unstable; urgency=medium

  [ Dmitry Yu Okunev ]
  * Added support of control socket

  [ Barak A. Pearlmutter ]
  * New upstream version
  * Bump standards version
  * Remove -*- makefile -*- in debian/rules: defaults to GNUmakefile-mode
  * Must generate library: ./configure --enable-socket
  * Fold clsync-dev into libclsync-dev
  * Address lintian issues
    - Add appropriate per-binary-package sections
    - Remove extra LICENSE file
    - Rename lib package w/ .so suffix "0"
    - Expand and diversify package descriptions.
  * Add dependency of lib-dev package upon lib package
  * Print list of debian/tmp files not installed into any binary package
  * Install some stray uninstalled files
    - the actual clsync executable (!)
    - its man page
    - some *.h header files
  * Fold small (14k) clsync-doc package into clsync

 -- Barak A. Pearlmutter <bap@debian.org>  Fri, 06 Jun 2014 09:44:28 +0100

clsync (0.2.1-1) unstable; urgency=low

  * New upstream version

 -- Barak A. Pearlmutter <bap@debian.org>  Thu, 24 Oct 2013 17:14:15 +0100

clsync (0.1-2) unstable; urgency=low

  * Tweak debian/watch to ignore debian releases

 -- Barak A. Pearlmutter <bap@debian.org>  Fri, 11 Oct 2013 10:45:20 +0100

clsync (0.1-1) unstable; urgency=low

  * Initial release (Closes: #718769 )

 -- Barak A. Pearlmutter <bap@debian.org>  Sat, 07 Sep 2013 12:16:00 +0100


ctx.h
/*
    clsync - file tree sync utility based on inotify

    Copyright (C) 2013  Dmitry Yu Okunev <dyokunev@ut.mephi.ru> 0x8E30679C

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#ifndef __CLSYNC_CTX_H
#define __CLSYNC_CTX_H

#include <regex.h>
#ifdef CAPABILITIES_SUPPORT
#	include <sys/capability.h>	// __u32
#endif

#define	MAX_BLOCKTHREADS	(1<<4)

#define register_blockthread(thread) {\
		critical_on (ctx_p->blockthread_count >= MAX_BLOCKTHREADS);\
		ctx_p->blockthread[ctx_p->blockthread_count++] = pthread_self();\
		debug(3, "register_blockthread(): ctx_p->blockthread_count -> %i", ctx_p->blockthread_count);\
	}

#define OPTION_FLAGS		(1<<10)
#define OPTION_LONGOPTONLY	(1<<9)
#define OPTION_CONFIGONLY	(1<<8)
#define NOTOPTION		(3<<8)
enum flags_enum {
	WATCHDIR	= 'W',
	SYNCHANDLER	= 'S',
	RULESFILE	= 'R',
	DESTDIR		= 'D',
	SOCKETPATH	= 's',

	HELP		= 'h',
	CONFIGFILE	= 'H',
	CONFIGBLOCK	= 'K',
	BACKGROUND	= 'b',
	UID		= 'u',
	GID		= 'g',
	CAP_PRESERVE	= 'C',
	THREADING	= 'p',
	RETRIES		= 'r',
	OUTPUT_METHOD	= 'Y',
	EXCLUDEMOUNTPOINTS = 'X',
	PIDFILE		= 'z',
	CLUSTERIFACE	= 'c',
	CLUSTERMCASTIPADDR = 'm',
	CLUSTERMCASTIPPORT = 'P',
	CLUSTERTIMEOUT	= 'G',
	CLUSTERNODENAME = 'n',
	CLUSTERHDLMIN	= 'o',
	CLUSTERHDLMAX	= 'O',
	DELAY		= 't',
	BFILEDELAY	= 'T',
	SYNCDELAY	= 'w',
	BFILETHRESHOLD	= 'B',
	DEBUG		= 'd',
	QUIET		= 'q',
	VERBOSE		= 'v',
	OUTLISTSDIR	= 'L',
	AUTORULESW	= 'A',
	MODE		= 'M',
	IGNOREEXITCODE	= 'x',
	DONTUNLINK	= 'U',
	INITFULL	= 'F',
	SYNCTIMEOUT	= 'k',
	LABEL		= 'l',
	SHOW_VERSION	= 'V',

	HAVERECURSIVESYNC 	=  0 | OPTION_LONGOPTONLY,
	RSYNCINCLIMIT		=  1 | OPTION_LONGOPTONLY,
	RSYNCPREFERINCLUDE	=  2 | OPTION_LONGOPTONLY,
	SYNCLISTSIMPLIFY	=  3 | OPTION_LONGOPTONLY,
	ONEFILESYSTEM		=  4 | OPTION_LONGOPTONLY,
	STATUSFILE		=  5 | OPTION_LONGOPTONLY,
	SKIPINITSYNC		=  6 | OPTION_LONGOPTONLY,
	ONLYINITSYNC		=  7 | OPTION_LONGOPTONLY,
	EXITONNOEVENTS		=  8 | OPTION_LONGOPTONLY,
	STANDBYFILE		=  9 | OPTION_LONGOPTONLY,
	EXITHOOK		= 10 | OPTION_LONGOPTONLY,
	CLUSTERSDLMAX		= 11 | OPTION_LONGOPTONLY,
	PREEXITHOOK		= 12 | OPTION_LONGOPTONLY,
	SOCKETAUTH		= 13 | OPTION_LONGOPTONLY,
	SOCKETMOD		= 14 | OPTION_LONGOPTONLY,
	SOCKETOWN		= 15 | OPTION_LONGOPTONLY,
	MAXITERATIONS		= 16 | OPTION_LONGOPTONLY,
	IGNOREFAILURES		= 17 | OPTION_LONGOPTONLY,
	DUMPDIR			= 18 | OPTION_LONGOPTONLY,
	CONFIGBLOCKINHERITS	= 19 | OPTION_LONGOPTONLY,
	MONITOR			= 20 | OPTION_LONGOPTONLY,
	SYNCHANDLERARGS0	= 21 | OPTION_LONGOPTONLY,
	SYNCHANDLERARGS1	= 22 | OPTION_LONGOPTONLY,
	CUSTOMSIGNALS		= 23 | OPTION_LONGOPTONLY,
	CHROOT			= 24 | OPTION_LONGOPTONLY,
	MOUNTPOINTS		= 25 | OPTION_LONGOPTONLY,
	SPLITTING		= 26 | OPTION_LONGOPTONLY,
	SYNCHANDLERUID		= 27 | OPTION_LONGOPTONLY,
	SYNCHANDLERGID		= 28 | OPTION_LONGOPTONLY,
	CAPS_INHERIT		= 29 | OPTION_LONGOPTONLY,
	CHECK_EXECVP_ARGS	= 30 | OPTION_LONGOPTONLY,
	PIVOT_ROOT		= 31 | OPTION_LONGOPTONLY,
	DETACH_NETWORK		= 32 | OPTION_LONGOPTONLY,
	DETACH_MISCELLANEA	= 33 | OPTION_LONGOPTONLY,
	ADDPERMITTEDHOOKFILES	= 34 | OPTION_LONGOPTONLY,
	SECCOMP_FILTER		= 35 | OPTION_LONGOPTONLY,
	FORGET_PRIVTHREAD_INFO	= 36 | OPTION_LONGOPTONLY,
	SECURESPLITTING		= 37 | OPTION_LONGOPTONLY,
	FTS_EXPERIMENTAL_OPTIMIZATION = 38 | OPTION_LONGOPTONLY,
	FORBIDDEVICES		= 39 | OPTION_LONGOPTONLY,
	CG_GROUPNAME		= 40 | OPTION_LONGOPTONLY,
	PERMIT_MPROTECT		= 41 | OPTION_LONGOPTONLY,
	SHM_MPROTECT		= 42 | OPTION_LONGOPTONLY,
	MODSIGN			= 43 | OPTION_LONGOPTONLY,
	CANCEL_SYSCALLS		= 44 | OPTION_LONGOPTONLY,
	EXITONSYNCSKIP		= 45 | OPTION_LONGOPTONLY,
	DETACH_IPC		= 46 | OPTION_LONGOPTONLY,
	PRIVILEGEDUID		= 47 | OPTION_LONGOPTONLY,
	PRIVILEGEDGID		= 48 | OPTION_LONGOPTONLY,
	SOFTEXITSYNC		= 49 | OPTION_LONGOPTONLY,
};
typedef enum flags_enum flags_t;

enum detachnetwork_way {
	DN_OFF = 0,
	DN_NONPRIVILEGED,
	DN_EVERYWHERE,
};
typedef enum detachnetwork_way detachnetwork_way_t;

enum pivotroot_way {
	PW_OFF    = 0,
	PW_DIRECT,
	PW_AUTO,
	PW_AUTORO,
};
typedef enum pivotroot_way pivotroot_way_t;

enum capsinherit {
	CI_DONTTOUCH = 0,
	CI_PERMITTED,
	CI_CLSYNC,
	CI_EMPTY,
};
typedef enum capsinherit capsinherit_t;

enum mode_id {
	MODE_UNSET	= 0,
	MODE_SIMPLE,
	MODE_DIRECT,
	MODE_SHELL,
	MODE_RSYNCSHELL,
	MODE_RSYNCDIRECT,
	MODE_RSYNCSO,
	MODE_SO,
};
typedef enum mode_id mode_id_t;

enum queue_id {
	QUEUE_NORMAL,
	QUEUE_BIGFILE,
	QUEUE_INSTANT,
	QUEUE_LOCKWAIT,

	QUEUE_MAX,
	QUEUE_AUTO
};
typedef enum queue_id queue_id_t;

enum ruleactionsign_enum {
	RS_REJECT	= 0,
	RS_PERMIT	= 1
};
typedef enum ruleactionsign_enum ruleactionsign_t;

enum ruleaction_enum {
	RA_NONE			 = 0x00,
	RA_MONITOR		 = 0x01,
	RA_SYNC			 = 0x02,
	RA_WALK			 = 0x04,
	RA_ALL			 = 0x0f,
};
typedef enum ruleaction_enum ruleaction_t;

// signals (man 7 signal)
enum sigusr_enum {
	SIGUSR_THREAD_GC	= 10,
	SIGUSR_INITSYNC  	= 12,
	SIGUSR_BLOPINT		= 16,
	SIGUSR_DUMP		= 29,
};

struct rule {
	int		num;
	regex_t		expr;
	mode_t		objtype;
	ruleaction_t	perm;
	ruleaction_t	mask;
};
typedef struct rule rule_t;

struct queueinfo {
	unsigned int 	collectdelay;
	time_t		stime;
};
typedef struct queueinfo queueinfo_t;

struct api_functs {
	api_funct_init   init;
	api_funct_sync   sync;
	api_funct_rsync  rsync;
	api_funct_deinit deinit;
};
typedef struct api_functs api_functs_t;

struct notifyenginefuncts {
	int ( *wait ) ( struct ctx *ctx_p, struct indexes *indexes_p, struct timeval *tv_p );
	int ( *handle ) ( struct ctx *ctx_p, struct indexes *indexes_p );
	int ( *add_watch_dir ) ( struct ctx *ctx_p, struct indexes *indexes_p, const char *const accpath );
};

enum shflags {
	SHFL_NONE		= 0x00,
	SHFL_RSYNC_ARGS		= 0x01,
	SHFL_INCLUDE_LIST	= 0x02,
	SHFL_INCLUDE_LIST_PATH	= 0x04,
	SHFL_EXCLUDE_LIST_PATH	= 0x08,
};
typedef enum shflags shflags_t;

enum shargsid {
	SHARGS_PRIMARY = 0,
	SHARGS_INITIAL,
	SHARGS_MAX,
};

struct synchandler_args {
	char	*v[MAXARGUMENTS];
	int	 c;
	char	 isexpanded[MAXARGUMENTS];
};
typedef struct synchandler_args synchandler_args_t;

#define STATE_STARTING(state_p) (state_p == NULL)
enum state_enum {
	STATE_EXIT 	= 0,
	STATE_STARTING,
	STATE_RUNNING,
	STATE_SYNCHANDLER_ERR,
	STATE_REHASH,
	STATE_PREEXIT,
	STATE_LASTSYNC,
	STATE_TERM,
	STATE_THREAD_GC,
	STATE_INITSYNC,
	STATE_HOLDON,
	STATE_UNKNOWN
};
typedef enum state_enum state_t;

static char *const status_descr[] = {
	[STATE_EXIT]		= "exiting",
	[STATE_STARTING]	= "starting",
	[STATE_RUNNING]		= "running",
	[STATE_SYNCHANDLER_ERR]	= "synchandler error",
	[STATE_REHASH]		= "rehashing",
	[STATE_PREEXIT]		= "preexit",
	[STATE_LASTSYNC]	= "lastsync",
	[STATE_TERM]		= "terminating",
	[STATE_THREAD_GC]	= "thread gc",
	[STATE_INITSYNC]	= "initsync",
	[STATE_HOLDON]		= "hold on",
	NULL
};

enum stat_fields {
	STAT_FIELD_RESET		= 0x0000,
	STAT_FIELD_DEV			= 0x0001,
	STAT_FIELD_INO			= 0x0002,
	STAT_FIELD_MODE			= 0x0004,
	STAT_FIELD_NLINK		= 0x0008,
	STAT_FIELD_UID			= 0x0010,
	STAT_FIELD_GID			= 0x0020,
	STAT_FIELD_RDEV			= 0x0040,
	STAT_FIELD_SIZE			= 0x0080,
	STAT_FIELD_BLKSIZE		= 0x0100,
	STAT_FIELD_BLOCKS		= 0x0200,
	STAT_FIELD_ATIME		= 0x0400,
	STAT_FIELD_MTIME		= 0x0800,
	STAT_FIELD_CTIME		= 0x1000,

	STAT_FIELD_ALL			= 0x1ff7,
};

enum syscall_bitmask {
	CSC_RESET	= 0x00,
	CSC_MON_STAT	= 0x01,
};

#define CAP_PRESERVE_TRY (1<<16)

struct ctx {
#ifndef LIBCLSYNC
	volatile state_t state;
	pid_t  pid;
	char   pid_str[65];
	size_t pid_str_len;
	uid_t uid;
	gid_t gid;
	uid_t privileged_uid;
	gid_t privileged_gid;
	uid_t synchandler_uid;
	gid_t synchandler_gid;
#ifdef CAPABILITIES_SUPPORT
	__u32 caps;
#endif
	pid_t child_pid[MAXCHILDREN];	// Used only for non-pthread mode
	int   children;			// Used only for non-pthread mode
	uint32_t iteration_num;
	rule_t rules[MAXRULES];
	size_t rules_count;
	dev_t st_dev;
#endif
	char *flags_values_raw[OPTION_FLAGS];
	int flags[OPTION_FLAGS];
	int flags_set[OPTION_FLAGS];
#ifndef LIBCLSYNC
	char *config_path;
	const char *config_block;
	char *customsignal[MAXSIGNALNUM + 1];
	char *label;
	char *watchdir;
	char *pidfile;
	char *standbyfile;
	char *exithookfile;
	char *preexithookfile;
	char *destdir;
	char *destproto;
	char *watchdirwslash;
	char *destdirwslash;
	char *statusfile;
	char *socketpath;
	char *dump_path;
#ifdef CGROUP_SUPPORT
	char *cg_groupname;
#endif
	int socket;
	mode_t socketmod;
	uid_t  socketuid;
	gid_t  socketgid;
#ifdef CLUSTER_SUPPORT
	char *cluster_iface;
	char *cluster_mcastipaddr;
	char *cluster_nodename;
	uint32_t cluster_nodename_len;
	uint16_t cluster_mcastipport;
	uint16_t cluster_hash_dl_min;
	uint16_t cluster_hash_dl_max;
	uint16_t cluster_scan_dl_max;
	unsigned int cluster_timeout;
#endif
	size_t watchdirlen;
	size_t destdirlen;
	size_t watchdirsize;
	size_t destdirsize;
	size_t watchdirwslashsize;
	size_t destdirwslashsize;
	short int watchdir_dirlevel;
	char *handlerfpath;
	void *handler_handle;
	api_functs_t handler_funct;
	char  *rulfpath;
	size_t rulfpathsize;
	char *listoutdir;
	struct notifyenginefuncts notifyenginefunct;
	int retries;
	size_t bfilethreshold;
	unsigned int syncdelay;
	queueinfo_t _queues[QUEUE_MAX];	// TODO: remove this from here
	unsigned int rsyncinclimit;
	time_t synctime;
	unsigned int synctimeout;
	sigset_t *sigset;
	char isignoredexitcode[ ( 1 << 8 )];
	pthread_t blockthread[MAX_BLOCKTHREADS];
	size_t    blockthread_count;

	char *chroot_dir;

#ifdef CAPABILITIES_SUPPORT
	char *permitted_hookfile[MAXPERMITTEDHOOKFILES + 1];
	int   permitted_hookfiles;
#endif

#ifdef UNSHARE_SUPPORT
# ifdef GETMNTENT_SUPPORT
	char *mountpoint[MAXMOUNTPOINTS + 1];
	int   mountpoints;
# endif
#endif

	synchandler_args_t synchandler_args[SHARGS_MAX];
	shflags_t synchandler_argf;
#endif // ifndef LIBCLSYNC
	void *indexes_p;
	void *fsmondata;
};
typedef struct ctx ctx_t;

#endif

